File: packages/mcp/src/client/http.ts
import type { AuthClient } from "../auth/oidc.js";
import type { Config } from "../config.js";
import { yaaifFetch } from "./tls.js";

function isTenantBootstrapPath(path: string): boolean {
  return (
    path.startsWith("/api/users/me/tenants") ||
    path.startsWith("/api/rbac/me") ||
    path.startsWith("/api/users/me/last-tenant") ||
    path.startsWith("/api/users/me/active-tenant")
  );
}

export class ApiClient {
  constructor(
    readonly cfg: Config,
    private readonly auth: AuthClient,
  ) {}

  agentJSON<T = unknown>(method: string, path: string, body?: unknown): Promise<T> {
    return this.doJSON<T>(this.cfg.agentBaseUrl, method, path, body);
  }

  apiJSON<T = unknown>(method: string, path: string, body?: unknown): Promise<T> {
    return this.doJSON<T>(this.cfg.apiBaseUrl, method, path, body);
  }

  controlPlaneJSON<T = unknown>(method: string, path: string, body?: unknown): Promise<T> {
    return this.doJSON<T>(this.cfg.controlPlaneBaseUrl, method, path, body);
  }

  approvalJSON<T = unknown>(method: string, path: string, body?: unknown): Promise<T> {
    return this.doJSON<T>(this.cfg.approvalBaseUrl, method, path, body);
  }

  private async doJSON<T>(base: string, method: string, path: string, body?: unknown): Promise<T> {
    let token: string;
    let session: Awaited<ReturnType<AuthClient["accessToken"]>>["session"];
    try {
      ({ token, session } = await this.auth.accessToken());
    } catch (e) {
      const msg = String(e);
      if (msg.includes("reauth_required") || msg.includes("issuer_mismatch") || msg.includes("not authenticated")) {
        throw e;
      }
      throw e;
    }
    const tenantId = (session.tenant_id || this.cfg.defaultTenantId || "").trim();
    if (!tenantId && !isTenantBootstrapPath(path)) {
      throw new Error("tenant not set; call yaaif_set_tenant (name/slug/uuid) or yaaif_ensure_session");
    }
    const headers: Record<string, string> = {
      Authorization: `Bearer ${token}`,
      Accept: "application/json",
    };
    if (tenantId) headers["X-Tenant-ID"] = tenantId;
    let payload: string | undefined;
    if (body !== undefined) {
      headers["Content-Type"] = "application/json";
      payload = JSON.stringify(body);
    }
    const res = await yaaifFetch(`${base}${path.startsWith("/") ? path : `/${path}`}`, {
      method,
      headers,
      body: payload,
    });
    const text = await res.text();
    if (!res.ok) {
      throw new Error(`${method} ${path} failed (${res.status}): ${text.trim()}`);
    }
    if (!text.trim() || res.status === 204) return undefined as T;
    return JSON.parse(text) as T;
  }
}

File: mcp-server/internal/client/http.go
package client

import (
	"bytes"
	"context"
	"encoding/json"
	"fmt"
	"io"
	"net/http"
	"net/url"
	"strings"
	"time"

	"yaaif-cursor-mcp/internal/auth"
	"yaaif-cursor-mcp/internal/config"
)

type API struct {
	cfg    config.Config
	auth   *auth.Client
	http   *http.Client
}

func New(cfg config.Config, authClient *auth.Client) *API {
	return &API{
		cfg:  cfg,
		auth: authClient,
		http: &http.Client{Timeout: 120 * time.Second},
	}
}

func (a *API) Config() config.Config { return a.cfg }

type RequestOptions struct {
	Query url.Values
}

func (a *API) AgentJSON(ctx context.Context, method, path string, body any, out any) error {
	return a.doJSON(ctx, a.cfg.AgentBaseURL, method, path, body, out)
}

func (a *API) APIJSON(ctx context.Context, method, path string, body any, out any) error {
	return a.doJSON(ctx, a.cfg.APIBaseURL, method, path, body, out)
}

func (a *API) doJSON(ctx context.Context, base, method, path string, body any, out any) error {
	token, sess, err := a.auth.AccessToken(ctx)
	if err != nil {
		return err
	}
	tenantID := strings.TrimSpace(sess.TenantID)
	if tenantID == "" {
		tenantID = strings.TrimSpace(a.cfg.DefaultTenantID)
	}
	if tenantID == "" && !isTenantBootstrapPath(path) {
		return fmt.Errorf("tenant not set; call yaaif_set_tenant or set YAAIF_DEFAULT_TENANT_ID")
	}

	var reader io.Reader
	if body != nil {
		raw, err := json.Marshal(body)
		if err != nil {
			return err
		}
		reader = bytes.NewReader(raw)
	}

	full := strings.TrimRight(base, "/") + "/" + strings.TrimLeft(path, "/")
	req, err := http.NewRequestWithContext(ctx, method, full, reader)
	if err != nil {
		return err
	}
	req.Header.Set("Authorization", "Bearer "+token)
	req.Header.Set("Accept", "application/json")
	if body != nil {
		req.Header.Set("Content-Type", "application/json")
	}
	if tenantID != "" {
		req.Header.Set("X-Tenant-ID", tenantID)
	}

	resp, err := a.http.Do(req)
	if err != nil {
		return err
	}
	defer resp.Body.Close()
	respBody, err := io.ReadAll(resp.Body)
	if err != nil {
		return err
	}
	if resp.StatusCode >= 300 {
		return fmt.Errorf("%s %s failed (%d): %s", method, path, resp.StatusCode, strings.TrimSpace(string(respBody)))
	}
	if out == nil || resp.StatusCode == http.StatusNoContent || len(bytes.TrimSpace(respBody)) == 0 {
		return nil
	}
	if err := json.Unmarshal(respBody, out); err != nil {
		return fmt.Errorf("decode response for %s: %w; body=%s", path, err, string(respBody))
	}
	return nil
}

func isTenantBootstrapPath(path string) bool {
	p := strings.TrimSpace(path)
	return strings.HasPrefix(p, "/api/users/me/tenants") ||
		strings.HasPrefix(p, "/api/rbac/me") ||
		strings.HasPrefix(p, "/api/users/me/last-tenant") ||
		strings.HasPrefix(p, "/api/users/me/active-tenant")
}
